File: backend/app/services/text_templates.py
from __future__ import annotations

from typing import Any


# Built-in packs when project settings omit text_templates (parity with frontend defaults)
_BUILTIN_TEMPLATES: dict[str, dict[str, Any]] = {
    "bubble": {
        "font_stack": ["NotoSansThai"],
        "font_size": 60,
        "auto_font_size": True,
        "min_font_size": 6,
        "max_font_size": 96,
        "color_hex": "#111111",
        "stroke_color": "#ffffff",
        "stroke_width": 0,
        "bold": False,
        "italic": False,
        "text_align": "center",
        "text_direction": "horizontal",
        "balloon_type": "bubble",
        "line_height_ratio": 1.2,
        "letter_spacing": 0,
        "padding": {"top": 12, "right": 18, "bottom": 12, "left": 18},
        # semantic_tag = AI brace label; template id IS the semantic role
        "semantic_tag": "ตัวละครพูด",
        "name": "บทพูดทั่วไป",
    },
    "narration": {
        "font_stack": ["NotoSansThai"],
        "font_size": 52,
        "auto_font_size": True,
        "min_font_size": 6,
        "max_font_size": 84,
        "color_hex": "#111111",
        "stroke_color": "#ffffff",
        "stroke_width": 0,
        "bold": False,
        "italic": False,
        "text_align": "left",
        "text_direction": "horizontal",
        "balloon_type": "narrative",
        "line_height_ratio": 1.2,
        "letter_spacing": 0,
        "padding": {"top": 16, "right": 20, "bottom": 16, "left": 20},
        "semantic_tag": "คำบรรยาย",
        "name": "คำบรรยาย",
    },
    "emphasis": {
        "font_stack": ["NotoSansThai"],
        "font_size": 72,
        "auto_font_size": True,
        "min_font_size": 6,
        "max_font_size": 120,
        "color_hex": "#111111",
        "stroke_color": "#ffffff",
        "stroke_width": 1,
        "bold": True,
        "italic": False,
        "text_align": "center",
        "text_direction": "horizontal",
        "balloon_type": "bubble",
        "line_height_ratio": 1.2,
        "letter_spacing": 0,
        "padding": {"top": 10, "right": 14, "bottom": 10, "left": 14},
        "semantic_tag": "ตะโกน",
        "name": "ตะโกน / เน้นเสียง",
    },
    "sfx": {
        "font_stack": ["Tahoma"],
        "font_size": 84,
        "auto_font_size": True,
        "min_font_size": 6,
        "max_font_size": 152,
        "color_hex": "#ffffff",
        "stroke_color": "#111111",
        "stroke_width": 3,
        "bold": True,
        "italic": True,
        "text_align": "center",
        "text_direction": "horizontal",
        "balloon_type": "sfx",
        "line_height_ratio": 1.2,
        "letter_spacing": 0,
        "padding": {"top": 4, "right": 4, "bottom": 4, "left": 4},
        "semantic_tag": "เสียงเอฟเฟกต์",
        "name": "เสียงเอฟเฟกต์",
    },
    "thought": {
        "font_stack": ["NotoSansThai"],
        "font_size": 52,
        "auto_font_size": True,
        "min_font_size": 6,
        "max_font_size": 84,
        "color_hex": "#444444",
        "stroke_color": "#ffffff",
        "stroke_width": 0,
        "bold": False,
        "italic": False,
        "text_align": "center",
        "text_direction": "horizontal",
        "balloon_type": "bubble",
        "line_height_ratio": 1.2,
        "letter_spacing": 0,
        "padding": {"top": 12, "right": 18, "bottom": 12, "left": 18},
        "semantic_tag": "คิดในใจ",
        "name": "คิดในใจ",
    },
    "system": {
        "font_stack": ["NotoSansThai"],
        "font_size": 48,
        "auto_font_size": True,
        "min_font_size": 6,
        "max_font_size": 72,
        "color_hex": "#ffffff",
        "stroke_color": "#000000",
        "stroke_width": 1,
        "bold": True,
        "italic": False,
        "text_align": "center",
        "text_direction": "horizontal",
        "balloon_type": "narrative",
        "line_height_ratio": 1.2,
        "letter_spacing": 0,
        "padding": {"top": 10, "right": 14, "bottom": 10, "left": 14},
        "semantic_tag": "ระบบพูด",
        "name": "ระบบพูด",
    },
}


def list_semantic_roles(settings: dict[str, Any] | None = None) -> list[dict[str, Any]]:
    """Every Font Template is a Semantic Role. Add template ⇒ add role."""
    templates = resolve_text_templates(settings)
    roles: list[dict[str, Any]] = []
    for tid, tmpl in templates.items():
        if not isinstance(tmpl, dict):
            continue
        roles.append({
            "id": str(tid),
            "template_id": str(tid),
            "name": str(tmpl.get("name") or tid),
            "semantic_tag": str(tmpl.get("semantic_tag") or "").strip(),
            "balloon_type": str(tmpl.get("balloon_type") or "bubble"),
        })
    return roles


def resolve_text_templates(settings: dict[str, Any] | None) -> dict[str, dict[str, Any]]:
    settings = settings or {}
    templates = settings.get("text_templates")
    if isinstance(templates, dict) and templates:
        out: dict[str, dict[str, Any]] = {}
        for key, value in templates.items():
            if isinstance(value, dict):
                out[str(key)] = value
        if out:
            return out
    return dict(_BUILTIN_TEMPLATES)


def get_default_text_template(settings: dict[str, Any] | None) -> tuple[str | None, dict[str, Any] | None]:
    """
    Resolve default template from project settings only.
    Does not inject built-in packs — callers that need builtins use resolve_text_templates().
    """
    settings = settings or {}
    templates = settings.get("text_templates")
    if not isinstance(templates, dict) or not templates:
        return None, None
    template_id = str(settings.get("default_text_template_id") or "bubble")
    template = templates.get(template_id)
    if not isinstance(template, dict):
        template_id, template = next(
            ((str(key), value) for key, value in templates.items() if isinstance(value, dict)),
            (None, None),
        )
    return template_id, template


def apply_template_by_id(
    block: Any,
    template_id: str | None,
    settings: dict[str, Any] | None,
) -> bool:
    """Apply a named template onto the block. Returns True if template was found."""
    if not template_id:
        return False
    templates = resolve_text_templates(settings)
    template = templates.get(str(template_id))
    if not isinstance(template, dict):
        return False
    return _apply_template_fields(block, str(template_id), template)


def _apply_template_fields(block: Any, template_id: str, template: dict[str, Any]) -> bool:
    raw_font_stack = template.get("font_stack") or [template.get("font_family")]
    if isinstance(raw_font_stack, str):
        raw_font_stack = [raw_font_stack]
    font_stack = list(dict.fromkeys(
        str(value).strip() for value in (raw_font_stack or []) if str(value).strip() and str(value) != "None"
    ))
    font = font_stack[0] if font_stack else "Tahoma"
    block.font_family = font
    metadata = dict(getattr(block, "extra_metadata", None) or {})
    effect_sources: dict[str, str] = dict(metadata.get("effect_sources") or {})

    # 1. Color: Priority AI Vision > Font Template
    extracted_color = getattr(block, "color_hex", None) or metadata.get("detected_color_hex")
    if extracted_color and str(extracted_color).startswith("#") and str(extracted_color).lower() not in {"#000000", "#111111", ""}:
        block.color_hex = str(extracted_color)
        effect_sources["color"] = "ai_vision"
    else:
        block.color_hex = str(template.get("color_hex", getattr(block, "color_hex", "#000000")))
        effect_sources["color"] = "template"

    # 2. Bold / Italic: Priority AI Vision > Font Template
    if metadata.get("detected_bold") is not None:
        block.bold = bool(metadata["detected_bold"])
        effect_sources["bold"] = "ai_vision"
    else:
        block.bold = bool(template.get("bold", False))
        effect_sources["bold"] = "template"

    if metadata.get("detected_italic") is not None:
        block.italic = bool(metadata["detected_italic"])
        effect_sources["italic"] = "ai_vision"
    else:
        block.italic = bool(template.get("italic", False))
        effect_sources["italic"] = "template"

    block.text_align = str(template.get("text_align", "center"))
    block.text_direction = str(template.get("text_direction", "horizontal"))
    if template.get("font_size") is not None:
        block.font_size = float(template["font_size"])
    metadata.pop("manual_font_size", None)
    semantic_tag = str(template.get("semantic_tag") or "").strip()
    semantic_label = semantic_tag or str(template.get("name") or template_id)

    # 3. Stroke: Priority AI Vision > Font Template
    detected_stroke = metadata.get("stroke_color")
    detected_stroke_w = metadata.get("stroke_width")
    if detected_stroke and str(detected_stroke).startswith("#") and str(detected_stroke).lower() not in {"#000000", ""}:
        stroke_color = detected_stroke
        stroke_width = detected_stroke_w if (detected_stroke_w and detected_stroke_w > 0) else 2.0
        effect_sources["stroke"] = "ai_vision"
    else:
        stroke_color = template.get("stroke_color", "#ffffff")
        stroke_width = template.get("stroke_width", 0) if template.get("stroke_enabled", True) else 0
        effect_sources["stroke"] = "template"

    # 4. Glow / Shadows / Gradients: Priority AI Vision > Template
    glow_enabled = template.get("outline_glow_enabled", True)
    if metadata.get("detected_gradient"):
        effect_sources["gradient"] = "ai_vision"
    if metadata.get("detected_drop_shadow"):
        effect_sources["drop_shadow"] = "ai_vision"
    if metadata.get("detected_outer_glow"):
        effect_sources["outer_glow"] = "ai_vision"
    if metadata.get("detected_inner_shadow"):
        effect_sources["inner_shadow"] = "ai_vision"

    # Legacy project templates predate the flag and were always auto-sized.
    # An explicit false is the only way to opt into a fixed template size.
    auto_font_size = bool(template.get("auto_font_size", True))
    metadata.update({
        "text_template_id": template_id,
        # Font Template id == Semantic Role id
        "semantic_role": template_id,
        "semantic_role_label": semantic_label,
        "semantic_role_template_id": template_id,
        # Preset geometry hint only — not detector evidence for Style Judge
        "template_balloon_type": str(template.get("balloon_type", "bubble")),
        "font_size_mode": "auto" if auto_font_size else "fixed",
        "auto_font_size": auto_font_size,
        "preferred_font_size": block.font_size,
        # Preserve every client-preset fallback. The typesetter resolves the
        # first installed family while the block keeps the customer's order.
        "font_stack": font_stack or [font],
        "stroke_color": stroke_color,
        "stroke_width": stroke_width,
        "outline_glow_color": template.get("outline_glow_color", template.get("stroke_color", "#ffffff")),
        "outline_glow_radius": template.get("outline_glow_radius", 0) if glow_enabled else 0,
        "outline_glow_opacity": template.get("outline_glow_opacity", 0) if glow_enabled else 0,
        "line_height_ratio": template.get("line_height_ratio", 1.2),
        "letter_spacing": template.get("letter_spacing", 0),
        "tracking": template.get("letter_spacing", 0),
        "padding": template.get("padding") or {},
        "effect_sources": effect_sources,
        "color_source": effect_sources.get("color", "template"),
    })
    if template.get("min_font_size") is not None:
        metadata["min_font_size"] = template["min_font_size"]
    if template.get("max_font_size") is not None:
        metadata["max_font_size"] = template["max_font_size"]
    block.extra_metadata = metadata
    return True


def apply_default_text_template(block: Any, settings: dict[str, Any] | None) -> bool:
    template_id, template = get_default_text_template(settings)
    if not template:
        fallback_font = str((settings or {}).get("default_font_family") or "Tahoma")
        block.font_family = fallback_font
        metadata = dict(getattr(block, "extra_metadata", None) or {})
        metadata["font_stack"] = [fallback_font]
        block.extra_metadata = metadata
        return False
    return _apply_template_fields(block, str(template_id), template)

File: workspaces/v1.0.0/frontend/src/utils/textTemplates.ts
import type { Project, TextBlock } from '../stores/projectStore';
import type { GradientSpec } from './typesetting';

export type WorkspaceMode = 'ocr' | 'typeset';

export interface TextTemplate {
  id: string;
  name: string;
  semantic_tag: string;
  font_stack: string[];
  font_size: number;
  auto_font_size?: boolean;
  min_font_size: number;
  max_font_size: number;
  color_hex: string;
  stroke_color: string;
  stroke_width: number;
  stroke_enabled?: boolean;
  outline_glow_color?: string;
  outline_glow_radius?: number;
  outline_glow_opacity?: number;
  outline_glow_enabled?: boolean;
  fill_type?: 'solid' | 'gradient';
  gradient_color_start?: string;
  gradient_color_end?: string;
  gradient_angle_deg?: number;
  gradient?: GradientSpec;
  drop_shadow_enabled?: boolean;
  drop_shadow_color?: string;
  drop_shadow_blur?: number;
  drop_shadow_offset_x?: number;
  drop_shadow_offset_y?: number;
  drop_shadow_opacity?: number;
  bold: boolean;
  italic: boolean;
  text_align: TextBlock['text_align'];
  text_direction: TextBlock['text_direction'];
  balloon_type: TextBlock['balloon_type'];
  line_height_ratio: number;
  letter_spacing: number;
  anti_alias?: 'smooth' | 'crisp' | 'strong' | 'sharp' | 'none';
  padding: { top: number; right: number; bottom: number; left: number };
}

export const DEFAULT_FONT_STACK = ['NotoSansThai'];

const disabledGradient = (color: string): GradientSpec => ({
  enabled: false,
  type: 'linear',
  stops: [{ position: 0, color }, { position: 1, color }],
  angle_deg: 0,
  scale: 100,
  reverse: false,
  dither: true,
  opacity: 1,
  blend_mode: 'normal',
});

export const DEFAULT_TEXT_TEMPLATES: Record<string, TextTemplate> = {
  bubble: {
    id: 'bubble', name: 'บทพูดทั่วไป', semantic_tag: 'ตัวละครพูด', font_stack: ['Baijam', 'TH Baijam', 'Baijam Bold', ...DEFAULT_FONT_STACK],
    font_size: 60, auto_font_size: true, min_font_size: 6, max_font_size: 96,
    color_hex: '#111111', stroke_color: '#ffffff', stroke_width: 0,
    gradient: disabledGradient('#111111'),
    bold: false, italic: false, text_align: 'center', text_direction: 'horizontal',
    balloon_type: 'bubble', line_height_ratio: 1.2, letter_spacing: 0, anti_alias: 'sharp',
    padding: { top: 12, right: 18, bottom: 12, left: 18 },
  },
  narration: {
    id: 'narration', name: 'คำบรรยาย', semantic_tag: 'คำบรรยาย', font_stack: ['TF PHETAI', 'TF Phetai', 'TF Phentai', ...DEFAULT_FONT_STACK],
    font_size: 52, auto_font_size: true, min_font_size: 6, max_font_size: 84,
    color_hex: '#111111', stroke_color: '#ffffff', stroke_width: 0,
    gradient: disabledGradient('#111111'),
    bold: false, italic: false, text_align: 'left', text_direction: 'horizontal',
    balloon_type: 'narrative', line_height_ratio: 1.2, letter_spacing: 0, anti_alias: 'sharp',
    padding: { top: 16, right: 20, bottom: 16, left: 20 },
  },
  emphasis: {
    id: 'emphasis', name: 'ตะโกน / เน้นเสียง', semantic_tag: 'ตะโกน', font_stack: ['TF PHETAI', 'TF Phetai', 'TF Phentai', 'Layiji_TarMineTine1', ...DEFAULT_FONT_STACK],
    font_size: 72, auto_font_size: true, min_font_size: 6, max_font_size: 120,
    color_hex: '#111111', stroke_color: '#ffffff', stroke_width: 2,
    gradient: disabledGradient('#111111'),
    bold: true, italic: true, text_align: 'center', text_direction: 'horizontal',
    balloon_type: 'bubble', line_height_ratio: 1.2, letter_spacing: 0, anti_alias: 'sharp',
    padding: { top: 10, right: 14, bottom: 10, left: 14 },
  },
  shout: {
    id: 'shout', name: 'ตะโกน / เสียงดัง', semantic_tag: 'ตะโกน', font_stack: ['TF PHETAI', 'TF Phetai', 'TF Phentai', 'Layiji_TarMineTine1', ...DEFAULT_FONT_STACK],
    font_size: 72, auto_font_size: true, min_font_size: 6, max_font_size: 120,
    color_hex: '#111111', stroke_color: '#ffffff', stroke_width: 2,
    gradient: disabledGradient('#111111'),
    bold: true, italic: true, text_align: 'center', text_direction: 'horizontal',
    balloon_type: 'bubble', line_height_ratio: 1.2, letter_spacing: 0, anti_alias: 'sharp',
    padding: { top: 10, right: 14, bottom: 10, left: 14 },
  },
  narrative: {
    id: 'narrative', name: 'คำบรรยาย', semantic_tag: 'คำบรรยาย', font_stack: ['TF PHETAI', 'TF Phetai', 'TF Phentai', ...DEFAULT_FONT_STACK],
    font_size: 52, auto_font_size: true, min_font_size: 6, max_font_size: 84,
    color_hex: '#111111', stroke_color: '#ffffff', stroke_width: 0,
    gradient: disabledGradient('#111111'),
    bold: false, italic: false, text_align: 'left', text_direction: 'horizontal',
    balloon_type: 'narrative', line_height_ratio: 1.2, letter_spacing: 0, anti_alias: 'sharp',
    padding: { top: 16, right: 20, bottom: 16, left: 20 },
  },
  sfx: {
    id: 'sfx', name: 'เสียงเอฟเฟกต์', semantic_tag: 'เสียงเอฟเฟกต์', font_stack: ['iannnnn TIGER Black', 'iannnnn TIGER', 'Tahoma', 'Impact'],
    font_size: 84, auto_font_size: true, min_font_size: 6, max_font_size: 152,
    color_hex: '#ffffff', stroke_color: '#111111', stroke_width: 3,
    gradient: disabledGradient('#ffffff'),
    bold: true, italic: true, text_align: 'center', text_direction: 'horizontal',
    balloon_type: 'sfx', line_height_ratio: 1.2, letter_spacing: 0, anti_alias: 'sharp',
    padding: { top: 4, right: 4, bottom: 4, left: 4 },
  },
  thought: {
    id: 'thought', name: 'คิดในใจ', semantic_tag: 'คิดในใจ', font_stack: ['TF PHETAI', 'TF Phetai', 'TF Phentai', ...DEFAULT_FONT_STACK],
    font_size: 52, auto_font_size: true, min_font_size: 6, max_font_size: 84,
    color_hex: '#444444', stroke_color: '#ffffff', stroke_width: 0,
    gradient: disabledGradient('#444444'),
    bold: false, italic: false, text_align: 'center', text_direction: 'horizontal',
    balloon_type: 'bubble', line_height_ratio: 1.2, letter_spacing: 0, anti_alias: 'sharp',
    padding: { top: 12, right: 18, bottom: 12, left: 18 },
  },
  whisper: {
    id: 'whisper', name: 'กระซิบ / แผ่วเบา', semantic_tag: 'กระซิบ', font_stack: ['TF PHETAI', 'TF Phetai', 'TF Phentai', 'FC Muffin', ...DEFAULT_FONT_STACK],
    font_size: 46, auto_font_size: true, min_font_size: 6, max_font_size: 72,
    color_hex: '#555555', stroke_color: '#ffffff', stroke_width: 0,
    gradient: disabledGradient('#555555'),
    bold: false, italic: false, text_align: 'center', text_direction: 'horizontal',
    balloon_type: 'bubble', line_height_ratio: 1.2, letter_spacing: 0, anti_alias: 'sharp',
    padding: { top: 12, right: 18, bottom: 12, left: 18 },
  },
  system: {
    id: 'system', name: 'ระบบพูด', semantic_tag: 'ระบบพูด', font_stack: ['IrisUPC', 'IrisPC', 'IrisUPC BOLD', 'IrisUPC-Bold', ...DEFAULT_FONT_STACK],
    font_size: 48, auto_font_size: true, min_font_size: 6, max_font_size: 72,
    color_hex: '#ffffff', stroke_color: '#000000', stroke_width: 3,
    gradient: disabledGradient('#ffffff'),
    bold: true, italic: false, text_align: 'center', text_direction: 'horizontal',
    balloon_type: 'narrative', line_height_ratio: 1.2, letter_spacing: 0, anti_alias: 'sharp',
    padding: { top: 10, right: 14, bottom: 10, left: 14 },
  },
};

export const normalizeTextTemplates = (value: unknown): Record<string, TextTemplate> => {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return { ...DEFAULT_TEXT_TEMPLATES };
  }
  const templates: Record<string, TextTemplate> = {};
  for (const [key, raw] of Object.entries(value as Record<string, Partial<TextTemplate>>)) {
    if (!raw || typeof raw !== 'object') continue;
    const fallback = DEFAULT_TEXT_TEMPLATES[raw.id || key] || DEFAULT_TEXT_TEMPLATES.bubble;
    templates[key] = {
      ...fallback,
      ...raw,
      id: raw.id || key,
      name: raw.name || key,
      // A template deliberately owns exactly one installed font. Older projects
      // may still contain fallback stacks, so normalize them on read.
      font_stack: [(
        Array.isArray(raw.font_stack)
          ? raw.font_stack.find((font): font is string => typeof font === 'string' && font.trim().length > 0)
          : undefined
      ) || fallback.font_stack[0]],
      padding: { ...fallback.padding, ...(raw.padding || {}) },
      gradient: raw.gradient && typeof raw.gradient === 'object'
        ? {
          ...fallback.gradient,
          ...raw.gradient,
          stops: Array.isArray(raw.gradient.stops) && raw.gradient.stops.length >= 2
            ? raw.gradient.stops
            : (fallback.gradient?.stops || [{ position: 0, color: '#111111' }, { position: 1, color: '#ffffff' }]),
        }
        : fallback.gradient,
      semantic_tag: (raw.semantic_tag ?? fallback.semantic_tag ?? '').toString(),
      stroke_enabled: typeof raw.stroke_enabled === 'boolean'
        ? raw.stroke_enabled
        : Number(raw.stroke_width ?? fallback.stroke_width ?? 0) > 0,
      outline_glow_enabled: typeof raw.outline_glow_enabled === 'boolean'
        ? raw.outline_glow_enabled
        : Number(raw.outline_glow_radius ?? fallback.outline_glow_radius ?? 0) > 0
          && Number(raw.outline_glow_opacity ?? fallback.outline_glow_opacity ?? 0) > 0,
    };
  }
  if (Object.keys(templates).length === 0) {
    return { ...DEFAULT_TEXT_TEMPLATES };
  }
  return templates;
};

/** Ensure every DEFAULT_TEXT_TEMPLATES id exists; never drop user custom templates. */
export function mergeDefaultTextTemplates(
  stored: Record<string, TextTemplate>,
): Record<string, TextTemplate> {
  const merged: Record<string, TextTemplate> = { ...stored };
  for (const [id, def] of Object.entries(DEFAULT_TEXT_TEMPLATES)) {
    if (!merged[id]) {
      merged[id] = { ...def, padding: { ...def.padding }, font_stack: [...def.font_stack] };
    } else {
      // Backfill semantic_tag on old entries that lacked the field
      const cur = merged[id];
      if (!(cur.semantic_tag || '').trim() && (def.semantic_tag || '').trim()) {
        merged[id] = { ...cur, semantic_tag: def.semantic_tag };
      }
    }
  }
  return merged;
}

/** Semantic Role catalog = Font Templates (id is both). */
export function listSemanticRolesFromTemplates(
  templates: Record<string, TextTemplate>,
): Array<{ id: string; name: string; semantic_tag: string }> {
  return Object.values(templates).map((t) => ({
    id: t.id,
    name: t.name,
    semantic_tag: (t.semantic_tag || '').trim(),
  }));
}

/**
 * Resolve what the UI should show as the block's Role/Font Template.
 * Prefer text_template_id (what user applied); fall back to semantic_role / template name.
 */
export function resolveBlockTemplateRole(
  block: {
    font_family?: string;
    font_size?: number;
    extra_metadata?: Record<string, unknown> | null;
  },
  templates: Record<string, TextTemplate>,
): {
  templateId: string | null;
  roleLabel: string;
  fontLabel: string;
  semanticTag: string;
  template: TextTemplate | null;
} {
  const meta = block.extra_metadata || {};
  const templateId = String(
    meta.text_template_id
    || meta.semantic_role_template_id
    || meta.semantic_role
    || '',
  ).trim() || null;

  const template = templateId
    ? (templates[templateId]
      || Object.values(templates).find((t) => t.id === templateId)
      || null)
    : null;
  const roleLabel = template
    ? (template.semantic_tag || template.name || template.id)
    : String(meta.semantic_role_label || meta.semantic_role || '').trim() || '—';
  const fontLabel = template
    ? `${template.font_stack[0] || '?'} · ${template.font_size}px`
    : `${block.font_family || '?'} · ${block.font_size ?? '?'}px`;
  const semanticTag = template
    ? (template.semantic_tag || '').trim()
    : String(meta.semantic_role_label || '').trim();

  return { templateId, roleLabel, fontLabel, semanticTag, template };
}

export const resolveGlobalTextTemplates = (
  locallyStored: unknown,
  projectStored: unknown,
): Record<string, TextTemplate> => {
  // Prefer local global presets, else project, else defaults — always merge builtins.
  if (locallyStored && typeof locallyStored === 'object' && !Array.isArray(locallyStored)
    && Object.keys(locallyStored as object).length > 0) {
    return mergeDefaultTextTemplates(normalizeTextTemplates(locallyStored));
  }
  if (projectStored && typeof projectStored === 'object' && !Array.isArray(projectStored)
    && Object.keys(projectStored as object).length > 0) {
    return mergeDefaultTextTemplates(normalizeTextTemplates(projectStored));
  }
  return { ...DEFAULT_TEXT_TEMPLATES };
};

export const templateBlockFields = (
  template: TextTemplate,
  currentMetadata: Record<string, unknown> = {},
): Partial<TextBlock> => {
  const templateMetadata = { ...currentMetadata };
  delete templateMetadata.manual_font_size;
  delete templateMetadata.typesetting_spec;
  const isAuto = Boolean(template.auto_font_size || template.font_size === 0 || !template.font_size);
  const resolvedFontSize = isAuto
    ? (template.font_size || Number(currentMetadata.source_font_size) || 18)
    : template.font_size;
  return {
    font_family: template.font_stack[0],
    font_size: resolvedFontSize,
    color_hex: template.color_hex,
    bold: template.bold,
    italic: template.italic,
    text_align: template.text_align,
    text_direction: template.text_direction,
    extra_metadata: {
      ...templateMetadata,
      text_template_id: template.id,
      // Font Template id == Semantic Role id
      semantic_role: template.id,
      semantic_role_label: (template.semantic_tag || template.name || template.id).trim(),
      semantic_role_template_id: template.id,
      // Preset geometry hint only — not detector evidence for Style Judge
      template_balloon_type: template.balloon_type,
      font_size_mode: isAuto ? 'auto' : 'fixed',
      preferred_font_size: template.font_size || null,
      auto_font_size: isAuto,
      font_stack: template.font_stack,
      min_font_size: template.min_font_size || 6,
      max_font_size: template.max_font_size || 120,
      stroke_color: template.stroke_color,
      stroke_width: template.stroke_enabled === false ? 0 : template.stroke_width,
      outline_glow_color: template.outline_glow_color || template.stroke_color,
      outline_glow_radius: template.outline_glow_enabled === false
        ? 0
        : Number(template.outline_glow_radius || 0),
      outline_glow_opacity: template.outline_glow_enabled === false
        ? 0
        : Number(template.outline_glow_opacity || 0),
      line_height_ratio: template.line_height_ratio,
      letter_spacing: template.letter_spacing,
      tracking: template.letter_spacing,
      padding: template.padding,
      gradient: template.gradient,
    },
  };
};

export const buildTemplateReapplicationUpdates = (
  project: Pick<Project, 'pages'>,
  templates: Record<string, TextTemplate>,
): Array<{ blockId: string; data: Partial<TextBlock> }> => (
  project.pages.flatMap((page) => page.text_blocks.flatMap((block) => {
    const templateId = String(
      block.extra_metadata?.text_template_id
      || block.extra_metadata?.semantic_role_template_id
      || '',
    );
    const template = templates[templateId];
    if (!template) return [];
    return [{
      blockId: block.id,
      data: templateBlockFields(template, {
        ...(block.extra_metadata || {}),
        source_font_size: block.extra_metadata?.source_font_size ?? block.font_size,
      }),
    }];
  }))
);
